# Data Parsing and Cleansing
## Task 1 and 2
#### Student Name: Jason Yong
#### Student ID: 3555422

Date: 23/08/21

Version: 1.0

Environment: Python 3 and Jupyter notebook

Libraries used: please include the main libraries you used in your assignment, e.g.,:
* pandas
* xml.etree.ElementTree
* random
* datetime

## Introduction

For Task 1, the jobs xml file was explored and the overall structure of the file was anaylsed to get an idea of how the data was being stored. The data in the file was then extracted an read into a pandas dataframe which allowed further data cleansing and examination to be conducted.

In Task 2, the data in each of the columns were examined to explore the different cell values that were present. Missing values and invalid values (e.g. converting salary ranges to average values) were then cleaned and replaced with appropriate entries using pandas functions e.g. .replace() function. The columns were also converted to their appropriate data types and the formats of cell values were also checked and changed to their appropriate formats (e.g. parsing dates and decimal places).

Lastly, the cleaned dataframe and a list of all errors/changes made to the dataframe were exported to csv files.

## Importing libraries 

In [1]:
import xml.etree.ElementTree as etree  
import pandas as pd
import random as ran
from datetime import datetime

# Task 1. Parsing Data

### 1.1. Examining and loading data

In [2]:
# Loading the data

tree = etree.parse("./S3555422_dataset1.xml")  
tree

In [3]:
# Examining the structure of the data

root = tree.getroot()
root.tag

'JobAds'

In [4]:
# Determining how many'children' of the root there are

len(root)

107

Each child corresponds to a different source hence there are 107 different sources

In [5]:
# Determining all the tags that are in the dataset

headers = set([child.tag for child in root.iter()])
headers

{'Category',
 'Company',
 'ContractTime',
 'ContractType',
 'Date',
 'Id',
 'JobAds',
 'Location',
 'Row',
 'Salary',
 'Source',
 'SourceName',
 'Title'}

In [6]:
# Exploring the structure of the first 'child'

for child in root[0].iter():
    print(child.tag, child.attrib, child.text)

Source {} 

SourceName {} theitjobboard.co.uk
Row {} 

Id {} 66606329
Title {} ASPNET MVC Developer  Central Liverpool  ****
Location {} Liverpool
ContractTime {} permanent
Category {} IT Jobs
Salary {} 34000
Date {'Open': '20130511T120000', 'Close': '20130809T120000'} None
Row {} 

Id {} 66195313
Title {} Senior Manager, Planning And Reporting
Location {} Yorkshire and Humberside
ContractTime {} permanent
Category {} IT Jobs
Salary {} 55000
Date {'Open': '20120310T000000', 'Close': '20120409T000000'} None
Row {} 

Id {} 67803705
Title {} Deployment Engineer OR Data center Build engineer
Location {} West Malling
ContractTime {} permanent
Category {} IT Jobs
Salary {} 35000 To 40000
Date {'Open': '20121128T120000', 'Close': '20130127T120000'} None
Row {} 

Id {} 64796847
Title {} Experienced CakePHP Developer
Location {} Surrey
ContractTime {} permanent
Category {} IT Jobs
Salary {} 45000
Date {'Open': '20130426T120000', 'Close': '20130625T120000'} None
Row {} 

Id {} 65171378
Title {} 

Date {'Open': '20130808T150000', 'Close': '20131106T150000'} None
Row {} 

Id {} 65899843
Title {} Software Solutions Architect  Manchester  ****
Location {} Altrincham
ContractTime {} permanent
Category {} IT Jobs
Salary {} 45000
Date {'Open': '20120625T150000', 'Close': '20120824T150000'} None
Row {} 

Id {} 68632622
Title {} SENIOR 3RD LINE SYSTEMS ENGINEER  ECOMMERCE  VMWARE/SAN/DR
Location {} Hertfordshire
ContractTime {} permanent
Category {} IT Jobs
Salary {} 52500
Date {'Open': '20120718T150000', 'Close': '20120817T150000'} None
Row {} 

Id {} 68635953
Title {} IT Consultant Project Manager
Location {} Suffolk
ContractTime {} permanent
Category {} IT Jobs
Salary {} 47500
Date {'Open': '20130607T000000', 'Close': '20130806T000000'} None
Row {} 

Id {} 65682056
Title {} Energy Consultant
Location {} London
ContractTime {} permanent
Category {} IT Jobs
Salary {} 34000
Date {'Open': '20120902T120000', 'Close': '20121201T120000'} None
Row {} 

Id {} 65171487
Title {} Amazing Recruit

Date {'Open': '20131112T120000', 'Close': '20131212T120000'} None
Row {} 

Id {} 67803683
Title {} Embedded Software Developer
Location {} Oxfordshire
ContractTime {} permanent
Category {} IT Jobs
Salary {} 50000
Date {'Open': '20120303T150000', 'Close': '20120502T150000'} None
Row {} 

Id {} 68634513
Title {} Infrastructure Support Engineer
Location {} The City
ContractTime {} contract
Category {} IT Jobs
Salary {} 70800
Date {'Open': '20121011T000000', 'Close': '20121025T000000'} None
Row {} 

Id {} 64558754
Title {} x**** Senior DATA Analysts  Southampton/Portsmouth
Location {} Southampton
ContractTime {} permanent
Category {} IT Jobs
Salary {} 45500
Date {'Open': '20130825T150000', 'Close': '20131123T150000'} None
Row {} 

Id {} 65899679
Title {} Product Manager  Medical Industry  Gloucestershire
Location {} Gloucestershire
ContractTime {} permanent
Category {} IT Jobs
Salary {} 47500
Date {'Open': '20120528T150000', 'Close': '20120826T150000'} None
Row {} 

Id {} 65171297
Title {}

Id {} 66195207
Title {} ASPNET/MVC Developer  URGENT North Manchester  ****
Location {} Manchester
ContractTime {} permanent
Category {} IT Jobs
Salary {} 39500
Date {'Open': '20130420T120000', 'Close': '20130719T120000'} None
Row {} 

Id {} 68049573
Title {} Software Developer (PHP Developer, MySQL, Developer)
Location {} Poole
ContractTime {} permanent
Category {} IT Jobs
Salary {} 27000
Date {'Open': '20120603T150000', 'Close': '20120901T150000'} None
Row {} 

Id {} 65681866
Title {} Java developer  Lead, Java, spring mvc, jboss, Northampton
Location {} Northamptonshire
ContractTime {} permanent
Category {} IT Jobs
Salary {} 42500
Date {'Open': '20120725T120000', 'Close': '20121023T120000'} None
Row {} 

Id {} 68049452
Title {} Project Manager/Manufacturing Derbyshire ****k
Location {} Derbyshire
ContractTime {} permanent
Category {} IT Jobs
Salary {} 40000 - 45000
Date {'Open': '20120205T120000', 'Close': '20120505T120000'} None
Row {} 

Id {} 64071634
Title {} Senior Java Web Deve

ContractTime {} permanent
Category {} IT Jobs
Salary {} 50000
Date {'Open': '20120825T120000', 'Close': '20121024T120000'} None
Row {} 

Id {} 66434911
Title {} SQL Developer  Redhill
Location {} Redhill
ContractTime {} permanent
Category {} IT Jobs
Salary {} 25500
Date {'Open': '20120628T150000', 'Close': '20120728T150000'} None
Row {} 

Id {} 67244582
Title {} Software Developers (C, Linux) required in Bristol  URGENT
Location {} UK
ContractTime {} permanent
Category {} IT Jobs
Salary {} 15000
Date {'Open': '20120503T000000', 'Close': '20120801T000000'} None
Row {} 

Id {} 67803619
Title {} Automated Software Tester (Selenium)  York  Circa ****
Location {} York
ContractTime {} permanent
Category {} IT Jobs
Salary {} 35000
Date {'Open': '20131107T150000', 'Close': '20140205T150000'} None
Row {} 

Id {} 64796994
Title {} Digital Designer  Digital  Designer UX Design
Location {} The City
ContractTime {} contract
Category {} IT Jobs
Salary {} 84000
Date {'Open': '20131202T150000', 'Close

Id {} 65899666
Title {} Operations Manager  Building Services
Location {} Surrey
ContractTime {} permanent
Category {} IT Jobs
Salary {} 60000
Date {'Open': '20121102T150000', 'Close': '20121202T150000'} None
Row {} 

Id {} 68049666
Title {} iOS Developer ****k  Huge, exciting international projects
Location {} South West England
ContractTime {} permanent
Category {} IT Jobs
Salary {} 50000
Date {'Open': '20120920T000000', 'Close': '20121119T000000'} None
Row {} 

Id {} 68635017
Title {} NET Developer  CNET, ASPNET, SQL Server  Mid level
Location {} Bedfordshire
ContractTime {} permanent
Category {} IT Jobs
Salary {} 42500
Date {'Open': '20120323T150000', 'Close': '20120621T150000'} None
Row {} 

Id {} 65437052
Title {} SAS  MI  PROGRAMMER  BASE  MACROS
Location {} North East England
ContractTime {} contract
Category {} IT Jobs
Salary {} 87000
Date {'Open': '20120312T000000', 'Close': '20120511T000000'} None
Row {} 

Id {} 64796941
Title {} Senior VBNet Developer Bath ****  ****
Locati

Given that each 'child' starts with the source, it means that each child represents a different source hence there are job ads from 107 different sources.

The only that has no text associated with it is the date. However, date has attributes which determine when the job ad is open and closed following the format YYYYMMDDTHHMMSS where T would denote the split between the date and time.

In [7]:
# Exploring the number of rows of information that are made available for the jobs

job_ad_length = []
tags = []
for child in root:
    for row in child:
        if row.tag != "SourceName":
            job_ad_length.append(len(row))
            for el in row:
                tags.append(el.tag)
                
set(job_ad_length)


{5, 6, 7, 8, 9}

In [8]:
set(tags)

{'Category',
 'Company',
 'ContractTime',
 'ContractType',
 'Date',
 'Id',
 'Location',
 'Salary',
 'Title'}

From this, the number of rows of information for each job in the data set varies from 5-9 rows of information. The pieces of information that are potentially included in each job ad are shown in the set of strings displayed in "tags"

In [9]:
# Exploring how many jobs there are for each source

job_nos = {}
for child in root:
    job_nos[child[0].text] = len(child) - 1
    
job_nos

{'theitjobboard.co.uk': 813,
 'justengineers.net': 384,
 'cwjobs.co.uk': 3038,
 'cv-library.co.uk': 7383,
 'totaljobs.com': 8371,
 'caterer.com': 2318,
 'gojobsearch.co.uk': 590,
 'britishjobsonthe.net': 332,
 'eFinancialCareers': 356,
 'jobs.guardian.co.uk': 1148,
 'fish4.co.uk': 1073,
 'careworx.co.uk': 1955,
 'contractjobs.com': 133,
 'jobs.catererandhotelkeeper.com': 1122,
 'ukstaffsearch.com': 434,
 'staffnurse.com': 2714,
 'planetrecruit.com': 616,
 'OilCareers.com': 49,
 'technojobs.co.uk': 460,
 'londonjobs.co.uk': 183,
 'jobsinrisk.com': 141,
 'professionalpensionsjobs.com': 77,
 'jobsite.co.uk': 2966,
 'jobs.electronicsweekly.com': 93,
 'hays.co.uk': 361,
 'Jobcentre Plus': 2057,
 'jobs.bighospitality.co.uk': 130,
 'Brand Republic Jobs': 400,
 'strike-jobs.co.uk': 358,
 'jobserve.com': 1569,
 'Jobs Ac': 91,
 'GAAPweb': 586,
 'nijobs.com': 176,
 'simplysalesjobs.co.uk': 64,
 'emedcareers.com': 276,
 'thecareerengineer.com': 141,
 'cityjobs.com': 300,
 'nijobfinder.co.uk': 134,

1 is deducted from each child as 1 row of each child corresponds to the name of the source

### 1.2 Parsing data into the required format

In [10]:
# Determining all the tags for the column values

from itertools import chain

set(chain.from_iterable([[child.tag for child in root[i].iter()] for i in range(0, len(root))]))

{'Category',
 'Company',
 'ContractTime',
 'ContractType',
 'Date',
 'Id',
 'Location',
 'Row',
 'Salary',
 'Source',
 'SourceName',
 'Title'}

In [11]:
dataDict = {}

sourcename = []
id = []
title = []
location = []
company = []
contracttype = []
contracttime = []
category = []
salary = []
date = []

for block in tree.iter(tag = "Source"):
    for child in block.findall("SourceName"):
        for i in range(0, job_nos[child.text]):
            sourcename.append(child.text)

for block in tree.iter(tag = "Row"):
    id.append([child.text for child in block.findall("Id")])
        
for block in tree.iter(tag = "Row"):
    title.append([child.text for child in block.findall("Title")])
        
for block in tree.iter(tag = "Row"):
    location.append([child.text for child in block.findall("Location")])
    
for block in tree.iter(tag = "Row"):
    company.append([child.text for child in block.findall("Company")])
    
for block in tree.iter(tag = "Row"):
    contracttype.append([child.text for child in block.findall("ContractType")])
    
for block in tree.iter(tag = "Row"):
    contracttime.append([child.text for child in block.findall("ContractTime")])
    
for block in tree.iter(tag = "Row"):
    category.append([child.text for child in block.findall("Category")])
    
for block in tree.iter(tag = "Row"):
    salary.append([child.text for child in block.findall("Salary")])
    
for block in tree.iter(tag = "Row"):
    date.append([child.attrib for child in block.findall("Date")])

    
# splitting the date column into open and close dates
opendate = []
closedate = []

for dates in date:
    opendate.append(dates[0]["Open"])
    closedate.append(dates[0]["Close"])

In [12]:
dataDict["Id"] = id
dataDict["Title"] = title
dataDict["Location"] = location
dataDict["Company"] = company
dataDict["ContractType"] = contracttype
dataDict["ContractTime"] = contracttime
dataDict["Category"] = category
dataDict["Salary"] = salary
dataDict["OpenDate"] = opendate
dataDict["CloseDate"] = closedate
dataDict["SourceName"] = sourcename

df = pd.DataFrame(dataDict)

# Task 2. Auditing and cleansing the loaded data

In [13]:
# creating the error list

itemlist = ['indexOfdf','Id', 'ColumnName','Original', 'Modified', 'ErrorType','Fixing']
errlist = pd.DataFrame(columns=itemlist)

In [14]:
# creating the function to append errors to the error list

def updateErrList(indexOfdf, Id, ColumnName, Orignal, Modified, ErrorType, Fixing):
    errItem=[indexOfdf, Id, ColumnName, Orignal, Modified, ErrorType, Fixing]
    errlist.loc[len(errlist)]=errItem

In [15]:
# Removing everything out of the lists and stripping any leading and trailing white spaces that may exist

col_names = df.columns.tolist()

for row in range(0, df.shape[0]):
    for col in range(0, len(col_names) - 3):
        if df.iloc[row, col] == []:
            df.iloc[row, col] = None
        else:
            df.iloc[row, col] = df.iloc[row, col][0].strip()
            
updateErrList("All", "All", "Id, Title, Location, Company, ContractType, ContractTime, Category, Salary", "All", "All", "String values contained within lists", "Remove the string out of the lists")
updateErrList("All", "All", "All", "All", "All","Leading and trailing white spaces", "Removing all leading and trailing white spaces with strip()")

In [16]:
# need to convert each column into their appropriate types

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50703 entries, 0 to 50702
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            50703 non-null  object
 1   Title         50703 non-null  object
 2   Location      50703 non-null  object
 3   Company       48817 non-null  object
 4   ContractType  34392 non-null  object
 5   ContractTime  45383 non-null  object
 6   Category      50703 non-null  object
 7   Salary        50194 non-null  object
 8   OpenDate      50703 non-null  object
 9   CloseDate     50703 non-null  object
 10  SourceName    50703 non-null  object
dtypes: object(11)
memory usage: 4.3+ MB


## Checking the Id column

In [17]:
# checking that all ID's are numeric and contain 8 digits using regex

num_mask = df["Id"].str.match('\d{8}') == True 
num_mask.sum()

50703

In [18]:
# converting the Id column to integer data type (int64)

df["Id"] = df["Id"].astype('int64')
df.info()

updateErrList('All', 'All', 'Id', 'All', 'All', 'Column dtype is object but should be integer','Converting dtype from "object" to "int64"')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50703 entries, 0 to 50702
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            50703 non-null  int64 
 1   Title         50703 non-null  object
 2   Location      50703 non-null  object
 3   Company       48817 non-null  object
 4   ContractType  34392 non-null  object
 5   ContractTime  45383 non-null  object
 6   Category      50703 non-null  object
 7   Salary        50194 non-null  object
 8   OpenDate      50703 non-null  object
 9   CloseDate     50703 non-null  object
 10  SourceName    50703 non-null  object
dtypes: int64(1), object(10)
memory usage: 4.3+ MB


In [19]:
# check for duplicates IDs

dup_mask = df["Id"].duplicated() == True
df.loc[dup_mask, :]

Id                                           Title  \
46534  69635711  Field Sales Account Manager  Mailing Equipment   

                Location                  Company ContractType ContractTime  \
46534  South East London  Collaborate Recruitment         None         None   

         Category Salary         OpenDate        CloseDate         SourceName  
46534  Sales Jobs  26000  20130209T120000  20130311T120000  salestarget.co.uk

Here we can see that there is one entry which has shares its Id with another job ad. Because each job ad needs to have a unique id, this value needs to be replaced with another 8 digit number.

In [20]:
index = df.loc[dup_mask, :].index

num = ran.randint(10000000,99999999) # randomly generate an 8 digit number
ids = df['Id'].tolist()

while num in ids: # check that the generated Id is not already being used
    num = ran.randint(10000000,99999999)

updateErrList(str(i), str(df.iloc[i,0]), 'Id', str(df.iloc[i,0]), str(num), 'Duplicate ID', 'Replace duplicate with unique 8 digit ID')

df.iloc[index[0], 0] = num 

## Checking the Title column

In [21]:
# convert to string

df["Title"] = df["Title"].astype('string')
df.info()

updateErrList('All', 'All', 'Title', 'All', 'All', 'Column dtype is object but should be string','Converting dtype from "object" to "string"')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50703 entries, 0 to 50702
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            50703 non-null  int64 
 1   Title         50703 non-null  string
 2   Location      50703 non-null  object
 3   Company       48817 non-null  object
 4   ContractType  34392 non-null  object
 5   ContractTime  45383 non-null  object
 6   Category      50703 non-null  object
 7   Salary        50194 non-null  object
 8   OpenDate      50703 non-null  object
 9   CloseDate     50703 non-null  object
 10  SourceName    50703 non-null  object
dtypes: int64(1), object(9), string(1)
memory usage: 4.3+ MB


## Checking the Location column

In [22]:
# convert to string

df["Location"] = df["Location"].astype("string")
df.info()

updateErrList('All', 'All', 'Location', 'All', 'All', 'Column dtype is object but should be string','Converting dtype from "object" to "string"')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50703 entries, 0 to 50702
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            50703 non-null  int64 
 1   Title         50703 non-null  string
 2   Location      50703 non-null  string
 3   Company       48817 non-null  object
 4   ContractType  34392 non-null  object
 5   ContractTime  45383 non-null  object
 6   Category      50703 non-null  object
 7   Salary        50194 non-null  object
 8   OpenDate      50703 non-null  object
 9   CloseDate     50703 non-null  object
 10  SourceName    50703 non-null  object
dtypes: int64(1), object(8), string(2)
memory usage: 4.3+ MB


In [23]:
# check strings that do no start with a capital letter

mask = df['Location'].str.match(r'^[a-z]') == True

index = df.loc[mask,:].index

for i in index:
    og = df.iloc[i,2]
    df.iloc[i,2] = og.capitalize()
    mod = df.iloc[i,2]
    
    updateErrList(str(i), str(df.iloc[i,0]), 'Location', og, mod, 'Location is not capitalised','Capitalise the first letter of the location')

In [24]:
# checking strings that are all caps

mask = df['Location'].str.match(r'^[A-Z][A-Z]+') == True

locs = list(set(df.loc[mask,['Location']]['Location'].tolist()))
locs

['BRISTOL', 'ABERDEEN', 'HAMpshire', 'UK']

Here we can see that there are 3 entries which have the wrong string structure. UK does not count as that is a legitimate entry

In [25]:
# removing UK from the list of wrong locs

locs.remove('UK')

# Modifying the location names to have the correct string structure

for loc in locs:
    mask = df['Location'] == loc
    index = df.loc[mask,:].index
    for i in index:
        og = df.iloc[i,2]
        df.iloc[i,2] = og.capitalize()
        mod = df.iloc[i,2]
    
        updateErrList(str(i), str(df.iloc[i,0]), 'Location', og, mod, 'Location string contains random capital letters','Capitalise the first letter of the location and lower case everthing else')

In [26]:
# checking all the entries in the column

dict(df['Location'].value_counts())

{'UK': 7624,
 'London': 6610,
 'South East London': 2687,
 'The City': 1129,
 'Central London': 830,
 'Manchester': 711,
 'Leeds': 665,
 'Surrey': 606,
 'Birmingham': 593,
 'Reading': 529,
 'West Midlands': 519,
 'Bristol': 439,
 'Berkshire': 406,
 'East Sheen': 378,
 'Nottingham': 373,
 'Oxford': 366,
 'Sheffield': 358,
 'Hampshire': 355,
 'Milton Keynes': 339,
 'Cambridge': 330,
 'Newcastle Upon Tyne': 327,
 'Kent': 320,
 'Guildford': 319,
 'Leicester': 284,
 'Hertfordshire': 283,
 'Essex': 258,
 'Aberdeen': 256,
 'Liverpool': 251,
 'Oxfordshire': 247,
 'West Yorkshire': 246,
 'Cheshire': 229,
 'Basingstoke': 217,
 'North West London': 212,
 'Glasgow': 212,
 'Northampton': 211,
 'Slough': 207,
 'West Sussex': 201,
 'Buckinghamshire': 199,
 'South West London': 194,
 'Lancashire': 194,
 'Gloucestershire': 193,
 'Bradford': 191,
 'Warwickshire': 190,
 'Derby': 190,
 'York': 189,
 'Belfast': 182,
 'Brighton': 181,
 'Chester': 180,
 'Swindon': 180,
 'Maidstone': 178,
 'Leicestershire': 1

From the dictionary above, we can see that there are 5 entries that only occur 1 or 2 times which are:
1. Manchaster
2. Leads
3. Cembridge
4. Livepool
5. Oxfords
    
These are obvious typos and can be easily modified to their correct names.

In [27]:
loc_dict = dict(df['Location'].value_counts())
loc_list = list(loc_dict.keys())
typos = loc_list[-5:]
typos

['Manchaster', 'Leads', 'Cembridge', 'Livepool', 'Oxfords']

In [28]:
# Fixing Manchaster

mask = df['Location'] == 'Manchaster'

index = df.loc[mask,:].index

for i in index:
    og = df.iloc[i,2]
    df.iloc[i,2] = 'Manchester'
    mod = df.iloc[i,2]
    
    updateErrList(str(i), str(df.iloc[i,0]), 'Location', og, mod, 'Location is spelt wrong','Replace location with correct spelling')

In [29]:
# Fixing Leads

mask = df['Location'] == 'Leads'

index = df.loc[mask,:].index

for i in index:
    og = df.iloc[i,2]
    df.iloc[i,2] = 'Leeds'
    mod = df.iloc[i,2]
    
    updateErrList(str(i), str(df.iloc[i,0]), 'Location', og, mod, 'Location is spelt wrong','Replace location with correct spelling')

In [30]:
# Fixing Cembridge

mask = df['Location'] == 'Cembridge'

index = df.loc[mask,:].index

for i in index:
    og = df.iloc[i,2]
    df.iloc[i,2] = 'Cambridge'
    mod = df.iloc[i,2]
    
    updateErrList(str(i), str(df.iloc[i,0]), 'Location', og, mod, 'Location is spelt wrong','Replace location with correct spelling')

In [31]:
# Fixing Livepool

mask = df['Location'] == 'Livepool'

index = df.loc[mask,:].index

for i in index:
    og = df.iloc[i,2]
    df.iloc[i,2] = 'Liverpool'
    mod = df.iloc[i,2]
    
    updateErrList(str(i), str(df.iloc[i,0]), 'Location', og, mod, 'Location is spelt wrong','Replace location with correct spelling')

In [32]:
# Fixing Oxfords

mask = df['Location'] == 'Oxfords'

index = df.loc[mask,:].index

for i in index:
    og = df.iloc[i,2]
    df.iloc[i,2] = 'Oxford'
    mod = df.iloc[i,2]
    
    updateErrList(str(i), str(df.iloc[i,0]), 'Location', og, mod, 'Location is spelt wrong','Replace location with correct spelling')

## Checking the Company column

In [33]:
# convert to string

df["Company"] = df["Company"].astype("string")
df.info()

updateErrList('All', 'All', 'Company', 'All', 'All', 'Column dtype is object but should be string','Converting dtype from "object" to "string"')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50703 entries, 0 to 50702
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            50703 non-null  int64 
 1   Title         50703 non-null  string
 2   Location      50703 non-null  string
 3   Company       48817 non-null  string
 4   ContractType  34392 non-null  object
 5   ContractTime  45383 non-null  object
 6   Category      50703 non-null  object
 7   Salary        50194 non-null  object
 8   OpenDate      50703 non-null  object
 9   CloseDate     50703 non-null  object
 10  SourceName    50703 non-null  object
dtypes: int64(1), object(7), string(3)
memory usage: 4.3+ MB


In [34]:
# creating a set of all entries

entries = set(df['Company'].tolist())

In [35]:
# checking for empty strings, "-", "N/A", and None

print('' in entries, '-' in entries, 'N/A' in entries, None in entries)

True True True False


In [36]:
# checking for empty entries i.e. <NA>

mask = df["Company"].isna() == True 

df.loc[mask, :]

Id                                              Title  \
0      66606329      ASPNET MVC Developer  Central Liverpool  ****   
1      66195313             Senior Manager, Planning And Reporting   
2      67803705  Deployment Engineer OR Data center Build engineer   
3      64796847                      Experienced CakePHP Developer   
4      65171378                  Oracle DBA Database Administrator   
...         ...                                                ...   
50692  68290740              Magento Webmaster (Fashion Ecommerce)   
50693  70229148                             PHP Developer  Bristol   
50694  69973924                                       SEO Engineer   
50695  69117113                              Research/Data Analyst   
50696  70229149                         CMS (Q****) content upload   

                       Location Company ContractType ContractTime Category  \
0                     Liverpool    <NA>         None    permanent  IT Jobs   
1      Yorkshire and Humberside    <NA>         None    permanent  IT Jobs   
2                  West Malling    <NA>         None    permanent  IT Jobs   
3                        Surrey    <NA>         None    permanent  IT Jobs   
4                         Wigan    <NA>         None    permanent  IT Jobs   
...                         ...     ...          ...          ...      ...   
50692                Manchester    <NA>         None     contract  IT Jobs   
50693                   Bristol    <NA>         None     contract  IT Jobs   
50694                  Nuneaton    <NA>         None     contract  IT Jobs   
50695                    London    <NA>         None     contract  IT Jobs   
50696             Milton Keynes    <NA>         None     contract  IT Jobs   

               Salary         OpenDate        CloseDate           SourceName  
0               34000  20130511T120000  20130809T120000  theitjobboard.co.uk  
1               55000  20120310T000000  20120409T000000  theitjobboard.co.uk  
2      35000 To 40000  20121128T120000  20130127T120000  theitjobboard.co.uk  
3               45000  20130426T120000  20130625T120000  theitjobboard.co.uk  
4               34000  20120417T150000  20120716T150000  theitjobboard.co.uk  
...               ...              ...              ...                  ...  
50692           23040  20131222T000000  20140121T000000            3desk.com  
50693           31200  20121217T120000  20121231T120000            3desk.com  
50694           24000  20130428T120000  20130528T120000            3desk.com  
50695            9600  20131129T120000  20131229T120000            3desk.com  
50696           36000  20120221T000000  20120421T000000            3desk.com  

[1886 rows x 11 columns]

From the previos 2 cells, we can see that there are empty strings, "N/A", and \<NA\> present in the column

In [37]:
# replacing '' with non-specified 
mask = df["Company"] == '' 

index = df.loc[mask, :].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'Company', '', 'non-specified', 'Company name is an empty string','Replace "" with "non-specified"')

df['Company'].replace('', 'non-specified', inplace = True)

In [38]:
# replacing '-' with non-specified 
mask = df["Company"] == '-' 

index = df.loc[mask, :].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'Company', '-', 'non-specified', 'Company name is "-"','Replace "-" with "non-specified"')

df['Company'].replace('-', 'non-specified', inplace = True)

In [39]:
# replacing 'N/A' with non-specified 

mask = df["Company"] == 'N/A' 

index = df.loc[mask, :].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'Company', 'N/A', 'non-specified', 'Company name is "N/A" (string)','Replace "N/A" with "non-specified"')

df['Company'].replace('N/A', 'non-specified', inplace = True)

In [40]:
# replacing <NA> with non-specified 

mask = df["Company"].isna() == True 

index = df.loc[mask, :].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'Company', '<NA>', 'non-specified', 'Company name is <NA>','Replace <NA> with "non-specified"')
    
df['Company'].fillna('non-specified', inplace = True)

## Checking the ContractType column

In [41]:
allowed_types = ["full_time", "part_time", "non-specified"]

# convert to string

df["ContractType"] = df["ContractType"].astype("string")
df.info()

updateErrList('All', 'All', 'ContractType', 'All', 'All', 'Column dtype is object but should be string','Converting dtype from "object" to "string"')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50703 entries, 0 to 50702
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            50703 non-null  int64 
 1   Title         50703 non-null  string
 2   Location      50703 non-null  string
 3   Company       50703 non-null  string
 4   ContractType  34392 non-null  string
 5   ContractTime  45383 non-null  object
 6   Category      50703 non-null  object
 7   Salary        50194 non-null  object
 8   OpenDate      50703 non-null  object
 9   CloseDate     50703 non-null  object
 10  SourceName    50703 non-null  object
dtypes: int64(1), object(6), string(4)
memory usage: 4.3+ MB


In [42]:
# checking the different entries that are present in the column

entries = set(df["ContractType"].tolist())
entries

{'', '-', <NA>, 'N/A', 'full_time', 'part_time'}

From the set of entries, we can see that there are 4 entries that do not belong which are '' (empty string), '-', 'N/A', and None (no value). These will need to be changed to 'non-specified'.

In [43]:
# replacing '' with non-specified 

mask = df["ContractType"] == '' 

index = df.loc[mask, :].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'ContractType', '', 'non-specified', 'Contract type is an empty string','Replace empty string with "non-specified"')

df["ContractType"].replace('', 'non-specified', inplace = True)

In [44]:
# replacing '' with non-specified 

mask = df["ContractType"] == '-' 

index = df.loc[mask, :].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'ContractType', '-', 'non-specified', 'Contract type is "-"','Replace "-" with "non-specified"')

df["ContractType"].replace('-', 'non-specified', inplace = True)

In [45]:
# replacing 'N/A' with non-specified 

mask = df["ContractType"] == 'N/A' 

index = df.loc[mask, :].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'ContractType', 'N/A', 'non-specified', 'Contract type is "N/A"','Replace "N/A" with "non-specified"')

df["ContractType"].replace('N/A', 'non-specified', inplace = True)

In [46]:
# replacing None with non-specified 

mask = df["ContractType"].isnull() == True 

index = df.loc[mask, :].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'ContractType', 'None type', 'non-specified', 'Contract type is "N/A"','Replace None type with "non-specified"')

df["ContractType"].fillna('non-specified', inplace = True)

In [47]:
# checking that the column now only contains the allowed types

set(allowed_types) == set(df["ContractType"].tolist())

True

In [48]:
set(df["ContractType"].tolist())

{'full_time', 'non-specified', 'part_time'}

The code above returns True hence all the contract types are 1 of the 3 valid types

## Checking the ContractTime column

In [49]:
# convert to string

df["ContractTime"] = df["ContractTime"].astype("string")
df.info()

updateErrList('All', 'All', 'ContractTime', 'All', 'All', 'Column dtype is object but should be string','Converting dtype from "object" to "string"')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50703 entries, 0 to 50702
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            50703 non-null  int64 
 1   Title         50703 non-null  string
 2   Location      50703 non-null  string
 3   Company       50703 non-null  string
 4   ContractType  50703 non-null  string
 5   ContractTime  45383 non-null  string
 6   Category      50703 non-null  object
 7   Salary        50194 non-null  object
 8   OpenDate      50703 non-null  object
 9   CloseDate     50703 non-null  object
 10  SourceName    50703 non-null  object
dtypes: int64(1), object(5), string(5)
memory usage: 4.3+ MB


In [50]:
# setting the allowed contract times to a list

allowed_times = ['contract', 'permanent', 'non-specified']

# checking the different entries that are present in the column

entries = set(df["ContractTime"].tolist())
entries

{'', '-', <NA>, 'N/A', 'contract', 'permanent'}

From the set of entries, we can see that there are 4 entries that do not belong which are '' (empty string), '-', 'N/A', and \<NA\> (no value). These will need to be changed to 'non-specified'.

In [51]:
# replacing '' with non-specified 

mask = df["ContractTime"] == '' 

index = df.loc[mask, :].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'ContractTime', '', 'non-specified', 'Contract time is an empty string','Replace empty string with "non-specified"')

df["ContractTime"].replace('', 'non-specified', inplace = True)

In [52]:
# replacing '-' with non-specified

mask = df["ContractTime"] == '-' 

index = df.loc[mask, :].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'ContractTime', '-', 'non-specified', 'Contract time is "-"','Replace "-" with "non-specified"')

df["ContractTime"].replace('-', 'non-specified', inplace = True)

In [53]:
# replacing <NA> with non-specified

mask = df["ContractTime"].isna() == True

index = df.loc[mask, :].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'ContractTime', '<NA>', 'non-specified', 'Contract time is missing a value','Replace missing value with "non-specified"')

df["ContractTime"].fillna('non-specified', inplace = True)

In [54]:
# replacing 'N/A' with non-specified

mask = df["ContractTime"] == 'N/A'

index = df.loc[mask, :].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'ContractTime', 'N/A', 'non-specified', 'Contract time is "N/A" (string)','Replace "N/A" with "non-specified"')

df["ContractTime"].replace('N/A', 'non-specified', inplace = True)

In [55]:
# checking all the entries in the column again to make sure that only allowed entries are present

set(allowed_times) == set(df["ContractTime"].tolist())

True

The code above returns True hence all contract times are 1 of the 3 valid times.

## Checking the Category column

In [56]:
# check that the categories are correct (i.e. one of the 8 allowed ones)

allowed_categories = ["IT Jobs", "Healthcare & Nursing Jobs", "Engineering Jobs", "Accounting & Finance Jobs", "Sales Jobs",
                     "Hospitality & Catering Jobs", "Teaching Jobs", "PR, Advertising & Marketing Jobs"]

# convert to string

df["Category"] = df["Category"].astype("string")
df.info()

updateErrList('All', 'All', 'Category', 'All', 'All', 'Column dtype is object but should be string','Converting dtype from "object" to "string"')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50703 entries, 0 to 50702
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            50703 non-null  int64 
 1   Title         50703 non-null  string
 2   Location      50703 non-null  string
 3   Company       50703 non-null  string
 4   ContractType  50703 non-null  string
 5   ContractTime  50703 non-null  string
 6   Category      50703 non-null  string
 7   Salary        50194 non-null  object
 8   OpenDate      50703 non-null  object
 9   CloseDate     50703 non-null  object
 10  SourceName    50703 non-null  object
dtypes: int64(1), object(4), string(6)
memory usage: 4.3+ MB


In [57]:
# checking the contents of the column

entries = set(df["Category"].tolist())

# comparing the set of column entries to the set of allowed categories

entries == set(allowed_categories) 

True

The logical statement returns True therefore all the entries in the column are valid and are part of the allowed categories

## Checking the Salary column

In [58]:
# converting to string to make it easier to work with

df["Salary"] = df["Salary"].astype("string")
df.info()

updateErrList('All', 'All', 'Salary', 'All', 'All', 'Column dtype is object but should be string','Converting dtype from "object" to "string"')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50703 entries, 0 to 50702
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Id            50703 non-null  int64 
 1   Title         50703 non-null  string
 2   Location      50703 non-null  string
 3   Company       50703 non-null  string
 4   ContractType  50703 non-null  string
 5   ContractTime  50703 non-null  string
 6   Category      50703 non-null  string
 7   Salary        50194 non-null  string
 8   OpenDate      50703 non-null  object
 9   CloseDate     50703 non-null  object
 10  SourceName    50703 non-null  object
dtypes: int64(1), object(3), string(7)
memory usage: 4.3+ MB


In [59]:
# check for any rows that contain alphabets

mask = df["Salary"].str.contains(r'[^\d{5,10}\.?\d+]') == True

df.loc[mask,:]

Id                                              Title  \
2      67803705  Deployment Engineer OR Data center Build engineer   
6      65171432   Senior PHP (LAMP) Developer  Oxford  Up to ****k   
164    64796544                                Web Developer (Net)   
172    62665255      Software / Web Developer  C  NET  ASPNET  MVC   
340    65437062                                Information Analyst   
...         ...                                                ...   
50598  68384687  Account Director / Group Account Director  Agency   
50608  68384710                            Senior Account Director   
50624  71073158            Senior Category Manager FMCG ( SL**** )   
50633  71073528          Industrial Engineer  Design for Six Sigma   
50639  71070610          Electronics /Electrical Engineer to ****k   

           Location                         Company   ContractType  \
2      West Malling                   non-specified  non-specified   
6            Oxford                   non-specified  non-specified   
164       Lowestoft                   non-specified  non-specified   
172         Suffolk                   non-specified  non-specified   
340           Leeds                   non-specified  non-specified   
...             ...                             ...            ...   
50598        Dorset                   non-specified  non-specified   
50608         Leeds                   non-specified  non-specified   
50624    Derbyshire                           JOBG8      full_time   
50633    Midlothian  Executive Recruitment Services      full_time   
50639     Berkshire          EasyWebRecruitment.com      full_time   

      ContractTime                          Category          Salary  \
2        permanent                           IT Jobs  35000 To 40000   
6        permanent                           IT Jobs      52500/Year   
164      permanent                           IT Jobs             32K   
172      permanent                           IT Jobs      34000/Year   
340       contract                           IT Jobs  29468 To 34468   
...            ...                               ...             ...   
50598    permanent  PR, Advertising & Marketing Jobs             N/A   
50608    permanent  PR, Advertising & Marketing Jobs             N/A   
50624    permanent                        Sales Jobs             N/A   
50633    permanent                  Engineering Jobs             N/A   
50639    permanent                  Engineering Jobs             N/A   

              OpenDate        CloseDate                 SourceName  
2      20121128T120000  20130127T120000        theitjobboard.co.uk  
6      20130719T150000  20130818T150000        theitjobboard.co.uk  
164    20120222T000000  20120323T000000        theitjobboard.co.uk  
172    20131214T000000  20140113T000000        theitjobboard.co.uk  
340    20121122T120000  20121206T120000        theitjobboard.co.uk  
...                ...              ...                        ...  
50598  20120510T120000  20120609T120000  simplymarketingjobs.co.uk  
50608  20130419T150000  20130718T150000  simplymarketingjobs.co.uk  
50624  20130321T120000  20130420T120000                  JobSearch  
50633  20130524T000000  20130723T000000                  JobSearch  
50639  20120226T150000  20120426T150000                  JobSearch  

[1684 rows x 11 columns]

From the output above we can see that there are over 1600 entries that contain a string, many of which are in different forms (e.g. some values are in ranges, some show yearly salary, and some show per hour) so each one will need to be dealt with individually.

In [60]:
# check for empty string

mask = df["Salary"] == ''

index = df.loc[mask,:].index.tolist()

len(index)

244

We can see that there are 244 entries where there is no salary value in the column. Since this column must contain a float, these rows will be removed from the dataframe.

In [61]:
# removing rows with empty string as salary

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'Salary', '', 'NA', 'Salary is an empty string','Entry deleted from dataframe')

df.drop(index, inplace = True)

In [62]:
# resetting the index and dropping the original index column

df.reset_index(inplace = True)
df.drop(columns = ['index'], inplace = True)

In [63]:
# check for 'N/A' - These we already know exists so we can just delete these entries

mask = df["Salary"] == 'N/A'

index = df.loc[mask,:].index.tolist()

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'Salary', 'N/A', 'NA', 'Salary is "N/A" string','Entry deleted from dataframe')

df.drop(index, inplace = True)

In [64]:
# resetting the index and dropping the original index column

df.reset_index(inplace = True)
df.drop(columns = ['index'], inplace = True)

In [65]:
# dealing with entries that are "-" - These we also know exist in the dataframe so we can just delete these entries

mask = df["Salary"] == "-"

index = df.loc[mask,:].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'Salary', '-', 'NA', 'Salary is "-" string','Entry deleted from dataframe')

df.drop(index, inplace = True)

In [66]:
# resetting the index and dropping the original index column

df.reset_index(inplace = True)
df.drop(columns = ['index'], inplace = True)

In [67]:
# checking for salary values that are 0

mask = df['Salary'] == "0"

df.loc[mask,:].shape

(516, 11)

Here we can see that there are 516 entries that have salary values equal to 0. These will be removed because it does not make sense to have a job that pays nothing

In [68]:
# removing entries with salary = 0

index = df.loc[mask,:].index.tolist()

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'Salary', '0', 'NA', 'Salary is 0','Entry deleted from dataframe')

df.drop(index, inplace = True)

In [69]:
# resetting the index and dropping the original index column

df.reset_index(inplace = True)
df.drop(columns = ['index'], inplace = True)

In [70]:
# dealing with entries with /Year or per Annum

mask = df['Salary'].str.contains(r'/Year|per Annum') == True

df.loc[mask,:].shape

(200, 11)

Here we can see that there are 200 entries that either have "per Annum" or "/Year" in the string. These will be replaced with just the numeric value

In [71]:
# checking to see that all entries with /Year and per Annum have their full numeric numbers e.g. not xxxK or xxxk

sum(df.loc[mask,["Salary"]]["Salary"].str.contains(r'[kK]') == True)

0

All entries "per Annum" or "/Year" show their complete numeric value.

In [72]:
# replacing all entries with /Year and per Anum to only show the numeric value

index = df.loc[mask,:].index

for i in index:
    og = df.iloc[i,7]
    df.iloc[[i],7] = df.iloc[[i],7].str.extract(r'(\d+)')
    mod = df.iloc[i,7]
    
    updateErrList(str(i), str(df.iloc[i,0]), 'Salary', og, mod, 'Salary is displayed in yearly sum (string)','Removed strings to only show numeric value')

In [73]:
# checking for entries that have their salary represented with abbreviations e.g. XXk or xxK

mask = df['Salary'].str.contains(r'[kK]') == True

df.loc[mask,:].shape

(50, 11)

We can see that there are 50 entries that contain salary values that are abbreviated

In [74]:
# replacing abbreviated values with their full value

index = df.loc[mask,:].index

for i in index:
    og = df.iloc[i,7]
    mod = df.iloc[[i],7].str.extract(r'(\d+)') + "000"
    df.iloc[[i],7] = mod

    updateErrList(str(i), str(df.iloc[i,0]), 'Salary', og, mod, 'Salary is displayed abbreviated form','Removed abbreviation and changed to full numeric form')

C:\Users\Jason\anaconda3\lib\site-packages\numpy\core\_asarray.py:102: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  return array(a, dtype, copy=False, order=order)


In [75]:
# dealing with salary entries that are given in ranges

mask = df["Salary"].str.contains(r'[Tt]o|-') == True

df.loc[mask,:].shape

(200, 11)

Here we can see that there are 200 entries where the values are displayed in ranges

In [76]:
# replacing range entries with average values

index = df.loc[mask,:].index

for i in index:
    og = df.iloc[i,7]
    lower = df.iloc[[i],7].str.extract(r'(\d{5,10}).+(\d{5,10})')[0]
    upper = df.iloc[[i],7].str.extract(r'(\d{5,10}).+(\d{5,10})')[1]
    mod = str(round((float(lower) + float(upper))/2,2))
    df.iloc[[i],7] = mod

    updateErrList(str(i), str(df.iloc[i,0]), 'Salary', og, mod, 'Salary is displayed in a range','Replace range with the average value')

In [77]:
# dealing with salary entries that are given in hourly rates
mask = df["Salary"].str.contains(r'per hour|p/h') == True

df.loc[mask,:].shape

(5, 11)

We can see that there are 5 entries that are given in hourly rates

In [78]:
# replacing hourly rate with yearly salary
# Assuming 52 work weeks, 5 days a week, and 8 hour work day

index = df.loc[mask,:].index

for i in index:
    og = df.iloc[i,7]
    rate = df.iloc[[i],7].str.extract(r'(\d+\.\d+)')[0]
    yearly = float(rate)*8*5*52
    mod = str(round(yearly))
    df.iloc[[i],7] = mod

    updateErrList(str(i), str(df.iloc[i,0]), 'Salary', og, mod, 'Salary is displayed in a hourly rate','Replace hourly rate with yearly salary')

In [79]:
# checking for null or None values

sum(df['Salary'].isnull() == True)

509

Here we can see that there are 509 entries where the values are missing. These will also be removed as null is not allowed as a value in the column.

In [80]:
# removing null and None

mask = df["Salary"].isnull() == True

index = df.loc[mask,:].index

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'Salary', 'null', 'NA', 'Salary is an empty string','Entry deleted from dataframe')

df.drop(index, inplace = True)

In [81]:
# resetting the index and dropping the original index column

df.reset_index(inplace = True)
df.drop(columns = ['index'], inplace = True)

In [82]:
# converting the column to float

df["Salary"] = df["Salary"].astype("float")

updateErrList('All', 'All', 'Salary', 'All', 'All', 'Column dtype is string but should be float','Converting dtype from string 2 float')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48205 entries, 0 to 48204
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Id            48205 non-null  int64  
 1   Title         48205 non-null  string 
 2   Location      48205 non-null  string 
 3   Company       48205 non-null  string 
 4   ContractType  48205 non-null  string 
 5   ContractTime  48205 non-null  string 
 6   Category      48205 non-null  string 
 7   Salary        48205 non-null  float64
 8   OpenDate      48205 non-null  object 
 9   CloseDate     48205 non-null  object 
 10  SourceName    48205 non-null  object 
dtypes: float64(1), int64(1), object(3), string(6)
memory usage: 4.0+ MB


In [83]:
# changing the format of the floats to 2 decimal places

pd.options.display.float_format = "{:,.2f}".format

updateErrList('All', 'All', 'Salary', 'All', 'All', 'Values are 1 decimal place','Formating the floats to display 2 decimals')

## Checking the OpenDate and CloseDate columns

In [84]:
# open and close dates need to be converted to datetime

df["OpenDate"] = df["OpenDate"].astype("string")
df["CloseDate"] = df["CloseDate"].astype("string")

updateErrList('All', 'All', 'OpenDate', 'All', 'All', 'Column dtype is object but should be string','Converting dtype from "object" to "string"')
updateErrList('All', 'All', 'CloseDate', 'All', 'All', 'Column dtype is object but should be string','Converting dtype from "object" to "string"')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48205 entries, 0 to 48204
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Id            48205 non-null  int64  
 1   Title         48205 non-null  string 
 2   Location      48205 non-null  string 
 3   Company       48205 non-null  string 
 4   ContractType  48205 non-null  string 
 5   ContractTime  48205 non-null  string 
 6   Category      48205 non-null  string 
 7   Salary        48205 non-null  float64
 8   OpenDate      48205 non-null  string 
 9   CloseDate     48205 non-null  string 
 10  SourceName    48205 non-null  object 
dtypes: float64(1), int64(1), object(1), string(8)
memory usage: 4.0+ MB


In [85]:
# check that all dates are in the format YYYYMMDDTHHMMSS

opendate_mask = df["OpenDate"].str.match(r'^\d{8}T\d{6}')
closedate_mask = df["CloseDate"].str.match(r'^\d{8}T\d{6}')
print(sum(opendate_mask), sum(closedate_mask))

48205 48205


The code above shows that all 50703 entries of the OpenDate and CloseDate columns are in the correct format

In [86]:
# Check Open year

open_years = df['OpenDate'].str[:4].tolist()
set(open_years)

{'2012', '2013'}

In [87]:
# Check Close year

close_years = df['CloseDate'].str[:4].tolist()
set(close_years)

{'2012', '2013', '2014'}

In [88]:
# Check Open month

open_months = df['OpenDate'].str[4:6].tolist()
set(open_months)

{'01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12', '29'}

There is one entry with a month value of 29 which does not make any sense

In [89]:
# Find the entry that as 29 for the month

mask = df['OpenDate'].str.match(r'^\d{4}29.+') == True

index = df.loc[mask,:].index.tolist()

updateErrList(str(index), str(df.iloc[index,0]), 'OpenDate', df.iloc[i,8], 'NA', 'Month value is 29 which does not make sense','Delete row as unable to tell when the true open date is')

# dropping the row because there it is unclear when the open date for the job is

df.drop(index, inplace = True)

In [90]:
# resetting the index and dropping the original index column

df.reset_index(inplace = True)
df.drop(columns = ['index'], inplace = True)

In [91]:
# Check Close month

close_months = df['CloseDate'].str[4:6].tolist()
set(close_months)

{'01', '02', '03', '04', '05', '06', '07', '08', '09', '10', '11', '12'}

In [92]:
# Check Open day

open_days = df['OpenDate'].str[6:8].tolist()
set(open_days)

{'01',
 '02',
 '03',
 '04',
 '05',
 '06',
 '07',
 '08',
 '09',
 '10',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '18',
 '19',
 '20',
 '21',
 '22',
 '23',
 '24',
 '25',
 '26',
 '27',
 '28',
 '29',
 '30',
 '31'}

In [93]:
# Check Close day

close_days = df['CloseDate'].str[6:8].tolist()
set(close_days)

{'01',
 '02',
 '03',
 '04',
 '05',
 '06',
 '07',
 '08',
 '09',
 '10',
 '11',
 '12',
 '13',
 '14',
 '15',
 '16',
 '17',
 '18',
 '19',
 '20',
 '21',
 '22',
 '23',
 '24',
 '25',
 '26',
 '27',
 '28',
 '29',
 '30',
 '31'}

In [94]:
# Check Open hour

open_hours = df['OpenDate'].str[9:11].tolist()
set(open_hours)

{'00', '12', '15'}

In [95]:
# Check Close hour

close_hours = df['CloseDate'].str[9:11].tolist()
set(close_hours)

{'00', '12', '15'}

In [96]:
# Check Open minute

open_minutes = df['OpenDate'].str[11:13].tolist()
set(open_minutes)

{'00'}

In [97]:
# Check Close minute

close_minutes = df['CloseDate'].str[11:13].tolist()
set(close_minutes)

{'00'}

In [98]:
# Check seconds

open_secs = df['OpenDate'].str[13:].tolist()
set(open_secs)

{'00'}

In [99]:
close_secs = df['OpenDate'].str[13:].tolist()
set(close_secs)

{'00'}

In [100]:
# Converting the OpenDate and CloseDate columns to datetime

df["OpenDate"] = df["OpenDate"].apply(lambda x: datetime.strptime(x, "%Y%m%dT%H%M%S").strftime("%Y-%m-%d %H:%M:%S"))
df["CloseDate"] = df["CloseDate"].apply(lambda x: datetime.strptime(x, "%Y%m%dT%H%M%S").strftime("%Y-%m-%d %H:%M:%S"))

updateErrList('All', 'All', 'OpenDate', 'All', 'All', 'Column dtype is string but should be datetime','Converting dtype from "string" to "datetime"')
updateErrList('All', 'All', 'CloseDate', 'All', 'All', 'Column dtype is string but should be datetime','Converting dtype from "string" to "datetime"')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48204 entries, 0 to 48203
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Id            48204 non-null  int64  
 1   Title         48204 non-null  string 
 2   Location      48204 non-null  string 
 3   Company       48204 non-null  string 
 4   ContractType  48204 non-null  string 
 5   ContractTime  48204 non-null  string 
 6   Category      48204 non-null  string 
 7   Salary        48204 non-null  float64
 8   OpenDate      48204 non-null  object 
 9   CloseDate     48204 non-null  object 
 10  SourceName    48204 non-null  object 
dtypes: float64(1), int64(1), object(3), string(6)
memory usage: 4.0+ MB


In [101]:
# Check that open date is not later than close date

mask = (df['OpenDate'] > df['CloseDate']) == True

df.loc[mask,:].shape

(5, 11)

The logical statement above returns 5 rows where the open date is later than the close date which cannot be possible. These rows will be deleted as it is not possible to determine when the true open date is

In [102]:
# removing the entries where the open date is later than the close date

index = df.loc[mask,:].index.tolist()

for i in index:
    updateErrList(str(i), str(df.iloc[i,0]), 'OpenDate', df.iloc[i,8], 'NA', 'Open date is later than the close date','Delete row as unable to tell when the true open date is')
    
df.drop(index, inplace = True)

In [103]:
# resetting the index and dropping the original index column

df.reset_index(inplace = True)
df.drop(columns = ['index'], inplace = True)

## Checking the SourceName column

In [104]:
# convert to string

df["SourceName"] = df["SourceName"].astype("string")

updateErrList('All', 'All', 'SourceName', 'All', 'All', 'Column dtype is object but should be string','Converting dtype from "object" to "string"')

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48199 entries, 0 to 48198
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Id            48199 non-null  int64  
 1   Title         48199 non-null  string 
 2   Location      48199 non-null  string 
 3   Company       48199 non-null  string 
 4   ContractType  48199 non-null  string 
 5   ContractTime  48199 non-null  string 
 6   Category      48199 non-null  string 
 7   Salary        48199 non-null  float64
 8   OpenDate      48199 non-null  object 
 9   CloseDate     48199 non-null  object 
 10  SourceName    48199 non-null  string 
dtypes: float64(1), int64(1), object(2), string(7)
memory usage: 4.0+ MB


## Saving data
Save the parsed and cleaned data as well as the error list, named as:
- \<student\_id\>_dataset1_solution.csv
- \<student\_id\>_errorlist.csv

In [105]:
# code to save output data

df.to_csv("S3555422_dataset1_solution.csv", index = False)
errlist.to_csv("S3555422_errorlist.csv", index = False)

## Summary

Overall, I found these tasks very time consuming and a little but tricky at the start. This was mainly due to trying to navigate and explore the structure of the xml file (which i'm not too good at doing). But after tinkering about with the code this was carried out fairly smoothly.

The other thing that I found tricky was trying to identify all the different errors that were present in the data and trying to figure out ways to deal with these errors (this part took up majority of my time).

As a whole, I really enjoyed doing these tasks and I learnt a lot about dealing with different datatypes, errors, and just using the pandas library and other python functions/libaries.